dulwich/porcelain.py

```
@@ -1753,6 +1755,136 @@ def update_head(repo, target, detached=False, new_branch=None):
+        index_entry[0] = (repo[b'HEAD'].commit_time, 0)  #ctime
+        index_entry[1] = (repo[b'HEAD'].commit_time, 0)  #mtime
+        index_entry[4] = tree_entry[0]  #mode
+        index_entry[7] = len(repo[tree_entry[1]].data)  #size
+        index_entry[8] = tree_entry[1]  #sha
+
+        index[path] = index_entry
+        index.write()
+
+
+def unstage_all(repo: Repo):
+    '''
+    unstage all file in the index
+    Args:
+        repo: dulwich Repo object
+        paths: a list of file to unstage
+    '''
+    index = repo.open_index()
+    tree_id = repo[b'HEAD'].tree
+    files_path = []
+
+    for entry in repo.object_store.iter_tree_contents(tree_id):
+        files_path.append(entry.path)
+    for entry in index:
+        files_path.append(entry[0])
+    # remove the same files
+    files_path = list(set(files_path))
+
+    unstage(repo, files_path)
+
+
+def reset_file(repo, file_path: str ,target: bytes = b'HEAD'):
+    '''
+    reset the file to specific commit or branch
+    Args:
+        repo: dulwich Repo object
+        file_path: file to reset
+        target: branch or commit or b'HEAD' to reset
+    '''
+    if target in branch_list(repo):
+        sha = repo.refs[b'refs/heads/' + target]
+    elif target == b'HEAD':
+        sha = repo.head()
+    else:
+        sha = target
+    tree = parse_tree(repo,treeish=sha)
+
+    def get_entry(repo, tree, file_path: str):
+        if os.path.split(file_path)[0] != '': # file_path with directory
+            par_dir_tree_id = get_entry(repo, tree, os.path.split(file_path)[0])[1]
+            tree = parse_tree(repo, treeish=par_dir_tree_id)
+            tree_id = tree[os.path.split(file_path)[1].encode()]
+            return tree_id
+        else:
+            try:
+                return tree[file_path.encode()]
+            except KeyError:
+                # untracked file not in tree.
+                print('file not in tree')
```

this should probably raise an exception rather than write to standard out (which may not be visible). Another option would be for the function to return the list of paths for files it actually changed.

---

I have considered about it. But I am confusing about which type of Error to raise.

---

Either KeyError or FileNotFoundError seem like they could be appropriate, or we could define a new exception type.

---

OK,It will raise KeyError in the new commit.